base/src/main/java/org/aya/core/visitor/Expander.java

```diff
@@ -105,9 +137,33 @@ record Normalizer(@NotNull TyckState state) implements Expander {
           assert tup.items().sizeGreaterThanOrEquals(ix) && ix > 0 : proj.toDoc(DistillerOptions.debug()).debugRender();
           yield apply(tup.items().get(ix - 1));
         }
+        case IntroTerm.PartEl el -> postPartial(el);
         default -> Expander.super.post(term);
       };
     }
+
+    public @NotNull Restr<Term> applyRestr(@NotNull Restr<Term> restr) {
+      return switch (restr.fmap(this)) {
+        case Restr.Vary<Term> vary -> CofThy.normalizeRestr(vary);
+        case Restr.Const<Term> c -> c;
+      };
+    }
+
+    private @NotNull IntroTerm.PartEl postPartial(@NotNull IntroTerm.PartEl el) {
+      return switch (el) {
+        case IntroTerm.SadPartEl par -> new IntroTerm.SadPartEl(par.u());
+        case IntroTerm.HappyPartEl par -> {
+          var clauses = MutableList.<Restr.Side<Term>>create();
+          for (var clause : par.clauses()) {
+            var u = clause.u();
+            if (CofThy.normalizeCof(clause.cof(), clauses, cofib -> new Restr.Side<>(cofib, u))) {
+              yield new IntroTerm.SadPartEl(u);
+            }
+          }
+          yield new IntroTerm.HappyPartEl(clauses.toImmutableSeq());
+        }
+      };
+    }
   }
 
   record WHNFer(@NotNull TyckState state) implements Expander {
```

Comment:
Should we also change the logic for `WHNFer`?
Or even unify most of that with `Normalizer`?

Reply:
Partial is already a whnf, I think it's fine.